### PR TITLE
Experiment (DO NOT MERGE) - Collapse ClusterHandler objects into PlatformEntity objects

### PR DESCRIPTION
**Pull request by dmulcahey**
**High-Level Summary**
This is a major architectural migration from **cluster-handler-centric runtime** to a **direct cluster-native runtime**.  
It is a systemic refactor that moves ownership of bind/report/init behavior and event wiring away from `zha/zigbee/cluster_handlers/*` into endpoint/device/entity flows.

---

**What Changed (Core Architecture)**

1. **Cluster handlers removed as a runtime abstraction**
- `zha/zigbee/cluster_handlers/` is deleted.
- Cluster policies/constants moved to new locations:
  - `zha/zigbee/cluster_policies.py`
  - `zha/zigbee/const.py`
- Event constant naming cleaned from `ZHA_CLUSTER_HANDLER_*` to `ZHA_CLUSTER_*` in `zha/application/const.py` (compat string payloads preserved where needed).

2. **Entity model switched from handlers to clusters**
- `ClusterHandlerMatch` replaced by `ClusterMatch` in `zha/application/platforms/__init__.py`.
- `PlatformEntity` now takes `clusters` and builds maps directly from in/out cluster collections.
- Entity metadata now exposes `clusters` (not `cluster_handlers`) through diagnostics info objects.
- Entities now declare/report lifecycle requirements directly:
  - `REPORT_CONFIG`
  - `ZCL_INIT_ATTRS`
- New merge logic added to combine per-entity requirements with cluster-level constraints, including quirk-direct precedence handling.

3. **Discovery rewritten to cluster-native matching**
- `zha/application/discovery.py` now matches entity classes via endpoint cluster names/IDs, not handler registries.
- Endpoint-level cluster claiming and lifecycle setup introduced (`claim_clusters(...)` path).
- Quirks v2 metadata now feeds entity/endpoint report-init requirements directly.

4. **Endpoint lifecycle is now the operational center**
- `zha/zigbee/endpoint.py` gained cluster-native bind/configure/init orchestration:
  - cluster claiming
  - bind decisions
  - reporting setup
  - attribute init reads
- Several side effects previously implemented in handlers now happen in endpoint/cluster listeners:
  - on/off client command state effects
  - OTA version state tracking
  - IAS Zone updates/enroll behavior
  - Identify trigger effects
  - DoorLock operation notification translation
- Compatibility payload fields were retained for some events to avoid breaking consumers during transition.

5. **Device orchestration changed materially**
- `zha/zigbee/device.py` now aggregates pending entity requirements and applies them before configure/init.
- Pending entity queue semantics changed (ordered map + key strategy for same-entity-id class variants).
- ZDO handler abstraction removed and replaced with direct endpoint-0/ZDO status handling.
- Diagnostics schema bumped to version 2 and field shape changed from handlers to clusters.

6. **Gateway/group robustness hardening added**
- `zha/application/gateway.py` introduces initialization generation tokens to prevent stale async completions from mutating current state.
- Startup polling task lifecycle and exception handling improved.
- Group reconcile operations now debounced.
- `zha/zigbee/group.py` adds lock-guarded reconciliation and more defensive subscription/entity cleanup.

7. **Helper utilities relocated**
- Exception wrapping/retry and safe command/write helpers moved into `zha/application/helpers.py`.
- Bindable policy checks now use cluster policy module, not handler registries.

---

**Platform Layer Migration Pattern**
Across platform implementations (examples: `zha/application/platforms/fan/__init__.py`, `zha/application/platforms/climate/__init__.py`, `zha/application/platforms/binary_sensor/__init__.py`, `zha/application/platforms/alarm_control_panel/__init__.py`, `zha/application/platforms/lock/__init__.py`, `zha/application/platforms/siren.py`):

- Constructor and internal state switched from `cluster_handlers` to `clusters`.
- Match declarations switched from `_cluster_handler_match` to `_cluster_match`.
- Event subscriptions now use direct cluster event types (`AttributeReadEvent`, `AttributeReportedEvent`, `AttributeUpdatedEvent`, `AttributeWrittenEvent`).
- Entity command/read/write code paths now call cluster methods/safe wrappers directly.
- Many entities gained explicit `REPORT_CONFIG` / `ZCL_INIT_ATTRS`.

This is mostly mechanical in simpler platforms, but in alarm/siren/lock it includes behavioral logic migration that is review-critical.

---

**Tests and Fixture Churn**
- `tests/test_cluster_handlers.py` removed.
- `tests/test_cluster_lifecycle.py` added for new endpoint/entity requirement aggregation and compatibility semantics.
- Large updates in `tests/test_gateway.py` and `tests/test_device.py` for async race hardening, pending entity handling, and cleanup behavior.
- Massive fixture migration in `tests/data/devices/`:
  - diagnostics schema version moved to 2
  - `cluster_handlers` fields replaced by `clusters`
  - 739 fixture JSONs touched

This explains the very large deletion count: fixture reshaping dominates diff volume.

---

**Commit Intent by Sequence (Review Narrative)**
- Early commits establish migration strategy/docs/scaffolding and naming cleanup.
- Mid commits perform systemic rename: handlers -> clusters, plus ZDO handling shift.
- Then platform-by-platform refactors migrate entity implementations.
- Final commits harden edge cases: async init races, queue behavior, listener cleanup, and add targeted tests.

---

**Review Focus Areas (Most Risk)**
1. `zha/zigbee/endpoint.py`  
   Risk: lifecycle ordering and parity with old handler side effects.
2. `zha/zigbee/device.py`  
   Risk: pending requirement merge semantics, ZDO transition, reconfigure flows.
3. `zha/application/discovery.py`  
   Risk: entity matching/claiming differences and quirks v2 metadata application.
4. `zha/application/gateway.py` + `zha/zigbee/group.py`  
   Risk: init/reconcile timing behavior and delayed convergence from debouncing.
5. Non-mechanical platform migrations (`alarm_control_panel`, `siren`, `lock`)  
   Risk: command semantics/state translation regressions.

---

**Suggested Review Plan**
1. Review the architectural pivot files first: `zha/zigbee/endpoint.py`, `zha/zigbee/device.py`, `zha/application/discovery.py`.
2. Review async hardening behavior: `zha/application/gateway.py`, `zha/zigbee/group.py`.
3. Review platform migrations by risk tier: `alarm_control_panel`, `siren`, `lock`, then broad mechanical platform files.
4. Review new lifecycle tests and fixture schema migration to confirm expected observability output shape.

### CODE REVIEW

File: zha/application/platforms/alarm_control_panel/__init__.py [hunk 1]
```diff
@@ -60,13 +63,13 @@ class AlarmControlPanel(PlatformEntity):
...
-    _cluster_handler_match = ClusterHandlerMatch(
-        client_cluster_handlers=frozenset({CLUSTER_HANDLER_IAS_ACE}),
+    _cluster_match = ClusterMatch(
+        out_clusters=frozenset({CLUSTER_IAS_ACE}),
     )
 
     def __init__(
         self,
-        cluster_handlers: list[ClusterHandler],
+        clusters: list[Any],
```
Comment (dmulcahey) [inline zha/application/platforms/alarm_control_panel/__init__.py]: typing

File: zha/application/platforms/climate/__init__.py [hunk 2]
```diff
@@ -689,15 +1325,17 @@ def current_temperature(self):
     @property
     def state(self) -> dict[str, Any]:
         """Get the state of the lock."""
-        thermostat = self._thermostat_cluster_handler
+        thermostat_system_mode = self._thermostat_get(
```
Comment (dmulcahey) [inline zha/application/platforms/climate/__init__.py]: name

File: zha/zigbee/device.py [hunk 3]
```diff
@@ -827,23 +876,24 @@ def extended_device_info(self) -> ExtendedDeviceInfo:
...
 
         await asyncio.gather(
             *(endpoint.async_configure() for endpoint in self._endpoints.values())
         )
 
         self.emit(
-            ZHA_CLUSTER_HANDLER_CFG_DONE,
-            ClusterHandlerConfigurationComplete(
+            ZHA_CLUSTER_CFG_DONE,
+            ClusterConfigurationComplete(
```
Comment (dmulcahey) [inline zha/zigbee/device.py:896]: hrm may break HA

File: zha/zigbee/group.py [hunk 4]
```diff
@@ -253,13 +264,51 @@ async def _maybe_update_group_members(self, event: EntityStateChangedEvent) -> N
...
+        if "info_object" in self.__dict__:
             delattr(self, "info_object")
-        if hasattr(self, "members"):
+        if "members" in self.__dict__:
             delattr(self, "members")
 
+    async def async_reconcile_discovered_entities(self) -> None:
+        """Reconcile currently registered group entities with discovery output."""
+        # Late-load discovery to avoid circular import with zha.application.discovery.
+        discovery: Any = importlib.import_module("zha.application.discovery")
```
Comment (dmulcahey) [inline zha/zigbee/group.py]: hrm... junk

File: zha/application/platforms/number/__init__.py [hunk 5]
```diff
@@ -208,29 +288,25 @@ class NumberConfigurationEntity(BaseNumber):
...
+        clusters: list[Cluster],
         endpoint: Endpoint,
         device: Device,
         **kwargs: Any,
     ) -> None:
         """Init this number configuration entity."""
-        self._cluster_handler: ClusterHandler = cluster_handlers[0]
-        super().__init__(cluster_handlers, endpoint, device, **kwargs)
+        super().__init__(clusters, endpoint, device, **kwargs)
+        self._cluster = clusters[0]
```
Comment (dmulcahey) [inline zha/application/platforms/number/__init__.py:303]: order flipped

File: zha/application/platforms/button/__init__.py [hunk 6]
```diff
@@ -60,19 +57,20 @@ class Button(PlatformEntity):
...
         device: Device,
         **kwargs: Any,
     ):
         """Initialize button."""
-        self._cluster_handler: ClusterHandler = cluster_handlers[0]
         if ENTITY_METADATA in kwargs:
             self._init_from_quirks_metadata(kwargs[ENTITY_METADATA])
-        super().__init__(cluster_handlers, endpoint, device, **kwargs)
+        super().__init__(clusters, endpoint, device, **kwargs)
+        self._cluster = clusters[0]
```
Comment (dmulcahey) [inline zha/application/platforms/button/__init__.py:73]: different sequencing

File: zha/application/platforms/binary_sensor/__init__.py [hunk 7]
```diff
@@ -68,29 +78,34 @@ class BinarySensor(PlatformEntity):
...
+        clusters: list[Cluster],
         endpoint: Endpoint,
         device: Device,
         **kwargs,
     ) -> None:
         """Initialize the ZHA binary sensor."""
-        self._cluster_handler = cluster_handlers[0]
-        super().__init__(cluster_handlers, endpoint, device, **kwargs)
+        super().__init__(clusters, endpoint, device, **kwargs)
+        self._cluster = clusters[0]
```
Comment (dmulcahey) [inline zha/application/platforms/binary_sensor/__init__.py:99]: order different

File: zha/application/platforms/binary_sensor/__init__.py [hunk 8]
```diff
@@ -125,30 +140,39 @@ def state(self) -> dict:
...
+        self._state = bool(event.value)
         self.maybe_emit_state_changed_event()
 
     async def async_update(self) -> None:
         """Attempt to retrieve on off state from the binary sensor."""
         await super().async_update()
-        attribute = getattr(self._cluster_handler, "value_attribute", "on_off")
         # this is a cached read to get the value for state mgt so there is no double read
-        attr_value = await self._cluster_handler.get_attribute_value(attribute)
+        attribute = self._attribute_name
```
Comment (dmulcahey) [inline zha/application/platforms/binary_sensor/__init__.py:173]: hrm

File: zha/application/platforms/binary_sensor/__init__.py [hunk 9]
```diff
@@ -342,18 +423,38 @@ def parse(value: bool | int) -> bool:
     async def async_update(self) -> None:
         """Attempt to retrieve on off state from the IAS Zone sensor."""
         await PlatformEntity.async_update(self)
+        attribute = self._attribute_name
```
Comment (dmulcahey) [inline zha/application/platforms/binary_sensor/__init__.py:573]: wonder why it kept creating vars for this...

File: zha/application/discovery.py [hunk 10]
```diff
@@ -326,72 +300,64 @@ def discover_quirks_v2_entities(device: Device) -> Iterator[PlatformEntity]:
...
+            # Bind is True by default, so only set to False if no reporting found.
+            # Quirks v2 entities are initialized last, so if the cluster wasn't
+            # already claimed, only v2 entities need it.
             if not reporting_found:
-                cluster_handler.BIND = False
+                endpoint.set_cluster_bind(cluster, False)
 
 
-def discover_entities_for_endpoint(endpoint: Endpoint) -> Iterator[PlatformEntity]:
+def discover_entities_for_endpoint(endpoint: Endpoint) -> Iterator[PlatformEntity]:  # noqa: C901
```
Comment (dmulcahey) [inline zha/application/discovery.py:355]: hrm

File: zha/zigbee/device.py [hunk 11]
```diff
@@ -173,13 +199,13 @@ class DeviceFirmwareInfoUpdatedEvent:
 
 
 @dataclass(kw_only=True, frozen=True)
-class ClusterHandlerConfigurationComplete:
-    """Event generated when all cluster handlers are configured."""
+class ClusterConfigurationComplete:
```
Comment (dmulcahey) [inline zha/zigbee/device.py:202]: wonder if this changes event shapes for HA

File: zha/zigbee/endpoint.py [hunk 12]
```diff
@@ -95,15 +253,35 @@ def id(self) -> int:
         """Return endpoint id."""
         return self._zigpy_endpoint.endpoint_id
 
-    @functools.cached_property
-    def cluster_handlers_by_name(self) -> dict[str, ClusterHandler]:
-        """Return cluster handlers indexed by name."""
-        return {ch.name: ch for ch in self._all_cluster_handlers.values()}
+    @property
```
Comment (dmulcahey) [inline zha/zigbee/endpoint.py:254]: hrm

File: zha/zigbee/endpoint.py [hunk 13]
```diff
@@ -95,15 +253,35 @@ def id(self) -> int:
...
+    def in_clusters_by_name(self) -> dict[str, Cluster]:
+        """Return input clusters indexed by cluster match name."""
+        return self._in_clusters_by_name
 
-    @functools.cached_property
-    def client_cluster_handlers_by_name(self) -> dict[str, ClientClusterHandler]:
-        """Return client cluster handlers indexed by name."""
-        return {ch.name: ch for ch in self._client_cluster_handlers.values()}
+    @property
```
Comment (dmulcahey) [inline zha/zigbee/endpoint.py:259]: hrm

File: zha/zigbee/endpoint.py [hunk 14]
```diff
@@ -135,127 +313,518 @@ def zigbee_signature(self) -> tuple[int, dict[str, Any]]:
...
-            gather = asyncio.gather
-        else:
-            gather = functools.partial(gather_with_limited_concurrency, max_concurrency)
+        client_commands = cluster.client_commands or {}
+        if command_id in client_commands:
+            return client_commands[command_id].name
+
+        return f"0x{command_id:02X}"
+
+    def handle_cluster_command(
```
Comment (dmulcahey) [inline zha/zigbee/endpoint.py:706]: ooh this is kinda ugly... wonder why these were all grouped together...

File: zha/zigbee/endpoint.py [hunk 15]
```diff
@@ -135,127 +313,518 @@ def zigbee_signature(self) -> tuple[int, dict[str, Any]]:
...
+                ATTRIBUTE_ID: event.attribute_id,
+                ATTRIBUTE_NAME: event.attribute_name or UNKNOWN,
+                ATTRIBUTE_VALUE: event.value,
+                VALUE: event.value,
+            },
+        )
 
-    async def _execute_handler_tasks(
-        self, func_name: str, *args: Any, max_concurrency: int | None = None
+    def _handle_smartthings_attribute_event(
```
Comment (dmulcahey) [inline zha/zigbee/endpoint.py]: hrm

File: zha/zigbee/endpoint.py (hunk 14 above)
Comment (dmulcahey) [reply zha/zigbee/endpoint.py:706]: I wonder if we may need to implement event entities to avoid remote logic from having to be handled like this... 🤔
